drivers/mspi/mspi_ambiq_ap3.c

```diff
@@ -1156,57 +1141,75 @@ static int mspi_dma_transceive(const struct device *controller,
+	if (ret) {
+		LOG_INST_ERR(cfg->log, "%u, failed to set sequence mode.", __LINE__);
+		ret = -EHOSTDOWN;
+		goto dma_err;
+	}
+
 	while (ctx->packets_left > 0) {
 		uint32_t packet_idx = ctx->xfer.num_packet - ctx->packets_left;
-		const struct mspi_xfer_packet *packet;
+		const struct mspi_xfer_packet *packet = &ctx->xfer.packets[packet_idx];
 
-		packet                     = &ctx->xfer.packets[packet_idx];
-		trans.ui8Priority          = ctx->xfer.priority;
 		trans.eDirection           = packet->dir;
-		trans.ui32TransferCount    = packet->num_bytes;
+		trans.ui16DeviceInstr      = packet->cmd;
 		trans.ui32DeviceAddress    = packet->address;
 		trans.ui32SRAMAddress      = (uint32_t)packet->data_buf;
-		trans.ui32PauseCondition   = 0;
-		trans.ui32StatusSetClr     = 0;
-
-		if (ctx->xfer.async) {
-
-			if (ctx->callback && packet->cb_mask == MSPI_BUS_XFER_COMPLETE_CB) {
-				ctx->callback_ctx->mspi_evt.evt_type = MSPI_BUS_XFER_COMPLETE;
-				ctx->callback_ctx->mspi_evt.evt_data.controller = controller;
-				ctx->callback_ctx->mspi_evt.evt_data.dev_id = data->ctx.owner;
-				ctx->callback_ctx->mspi_evt.evt_data.packet = packet;
-				ctx->callback_ctx->mspi_evt.evt_data.packet_idx = packet_idx;
-				ctx->callback_ctx->mspi_evt.evt_data.status = ~0;
-			}
-
-			am_hal_mspi_callback_t callback = NULL;
-
-			if (packet->cb_mask == MSPI_BUS_XFER_COMPLETE_CB) {
-				callback = (am_hal_mspi_callback_t)ctx->callback;
-			}
+		trans.ui32TransferCount    = packet->num_bytes;
+		trans.ui8Priority          = ctx->xfer.priority;
+		trans.ui16PacketIndex      = packet_idx;
 
-			ret = am_hal_mspi_nonblocking_transfer(data->mspiHandle, &trans, MSPI_DMA,
-							       callback, (void *)ctx->callback_ctx);
-		} else {
-			ret = am_hal_mspi_nonblocking_transfer(data->mspiHandle, &trans, MSPI_DMA,
-							       hal_mspi_callback,
-							       (void *)controller);
-		}
-		ctx->packets_left--;
+		ret = am_hal_mspi_cq_scatter_xfer(data->mspiHandle, &trans,
+						  hal_mspi_callback, (void *)controller);
 		if (ret) {
 			if (ret == AM_HAL_STATUS_OUT_OF_RANGE) {
+				LOG_INST_ERR(cfg->log, "%u, failed to transfer, cq out of memory.",
+					     __LINE__);
 				ret = -ENOMEM;
 			} else {
+				LOG_INST_ERR(cfg->log, "%u, failed to transfer, error %d",
+					     __LINE__, ret);
 				ret = -EIO;
+				am_hal_mspi_disable(data->mspiHandle);
+				am_hal_mspi_enable(data->mspiHandle);
 			}
 			goto dma_err;
 		}
+
+		ctx->packets_left--;
 	}
 
 	if (!ctx->xfer.async) {
-		while (ctx->packets_done < ctx->xfer.num_packet) {
+		uint32_t since = k_uptime_get_32();
+		uint32_t now = since;
+		uint32_t elapsed = 0;
+
+		while (ctx->packets_done < ctx->xfer.num_packet &&
+		       elapsed < ctx->xfer.timeout) {
 			k_busy_wait(10);
+			now = k_uptime_get_32();
+			if (now > since) {
+				elapsed = now - since;
+			} else {
+				elapsed = now + (uint32_t) - since;
+			}
+		}
+		if (elapsed >= ctx->xfer.timeout &&
+		    ctx->packets_done != ctx->xfer.num_packet) {
```

Comment:
is  `elapsed >= ctx->xfer.timeout` not true and `ctx->packets_done != ctx->xfer.num_packet` true possible?

Reply:
It is possible if there is a failure in hardware or software bug in hal. This allows capturing it.

But anyway it should be 
```suggestion
		if (elapsed >= ctx->xfer.timeout ||
		    ctx->packets_done != ctx->xfer.num_packet) {
```